server/http/filters/filters.go

```diff
@@ -93,14 +104,96 @@ func RequestID(next http.Handler) http.Handler {
 	})
 }
 
+// instrumentedResponseWriter wraps http.ResponseWriter, recording info needed for
+// Prometheus metrics.
+type instrumentedResponseWriter struct {
+	http.ResponseWriter
+
+	statusCode        int
+	responseSizeBytes int
+}
+
+func (w instrumentedResponseWriter) Header() http.Header {
+	return w.ResponseWriter.Header()
+}
+func (w instrumentedResponseWriter) Write(bytes []byte) (int, error) {
+	w.responseSizeBytes += len(bytes)
+	return w.ResponseWriter.Write(bytes)
+}
+func (w instrumentedResponseWriter) WriteHeader(statusCode int) {
+	w.statusCode = statusCode
+	w.ResponseWriter.WriteHeader(statusCode)
+}
+func (w instrumentedResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	return w.ResponseWriter.(http.Hijacker).Hijack()
+}
+func (w instrumentedResponseWriter) Flush() {
+	w.ResponseWriter.(http.Flusher).Flush()
+}
+
 func LogRequest(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		start := time.Now()
-		next.ServeHTTP(w, r)
-		log.LogHTTPRequest(r.Context(), r.URL.Path, time.Now().Sub(start), nil)
+		method := getMethod(r)
+		route := getRoute(r)
+		recordRequestMetrics(route, method)
+		ow := instrumentedResponseWriter{
+			ResponseWriter: w,
+		}
+		next.ServeHTTP(ow, r)
+		duration := time.Since(start)
+		log.LogHTTPRequest(r.Context(), r.URL.Path, duration, nil)
+		recordResponseMetrics(route, method, ow, duration)
```

Comment:
If you combined the record* funcs you can then do:

start := time.Now()
irw := instrumentedResponseWriter{ResponseWriter: w}
next.ServeHTTP(irw, r)
duration := time.Since(start)
log.LogHTTPRequest(r.Context(), r.URL.Path, duration, nil)
recordResponseMetrics(route(r), method(r), irw, duration)

Reply:
I'm logging requests separately from responses because if we were to see a bunch of requests logged but no responses logged, then that would point to a bug somewhere in the HTTP handler stack (ServeHTTP)